### ISSUE TITLE
False Negative: Non-serializable fields are not fully covered across type and declaration variants.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6533]
# False Negative: Non-serializable fields are not fully covered across type and declaration variants.

Version
7.22.0

## Checker
- Checker id: `category/java/errorprone.xml/NonSerializableClass`
- Checker description: Detects non-serializable fields in Serializable classes that are not handled by custom serialization methods or serialPersistentFields.

## Description of the false negative
`diff/output` marks 7 false negatives for `category/java/errorprone.xml/NonSerializableClass`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are suppression or annotations.

### `PosCase2.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Collection of non-serializable element type should be flagged.

```java
// Serializable class via inheritance with Collection of non-serializable element type should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.List;

class NonSerializableElement {}

public class PosCase2 implements Serializable {
    // Non-transient, non-static field of parameterized Collection with non-serializable element type
    private List<NonSerializableElement> items; // [REPORTED LINE]
}
```

### `PosCase3.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Collection of non-serializable element type and anonymous class not assigned to variable should be flagged.

```java
// Serializable class via inheritance with Collection of non-serializable element type and anonymous class not assigned to variable should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.List;

class NonSerializableElement2 {}

public class PosCase3 implements Serializable {
    private List<NonSerializableElement2> items; // [REPORTED LINE]
    
    public void method() {
        // Anonymous class not assigned to a variable
        new Serializable() {};
    }
}
```

### `PosCase4.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase4.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Collection of non-serializable element type and anonymous class assigned to variable without @SuppressWarnings should be flagged.

```java
// Serializable class via inheritance with Collection of non-serializable element type and anonymous class assigned to variable without @SuppressWarnings should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.Set;

class NonSerializableElement3 {}

public class PosCase4 implements Serializable {
    private Set<NonSerializableElement3> items; // [REPORTED LINE]
    
    public void method() {
        // Anonymous class instantiated and assigned to variable without @SuppressWarnings("serial")
        Serializable anonymous = new Serializable() {};
    }
}
```

### `PosCase5.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase5.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Map containing non-serializable value type should be flagged.

```java
// Serializable class via inheritance with Map containing non-serializable value type should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.Map;

class NonSerializableValue {}

public class PosCase5 implements Serializable {
    // Non-transient, non-static field of Map with non-serializable value type
    private Map<String, NonSerializableValue> map; // [REPORTED LINE]
}
```

### `PosCase6.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase6.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Map containing non-serializable key type and anonymous class assigned to variable without @SuppressWarnings should be flagged.

```java
// Serializable class via inheritance with Map containing non-serializable key type and anonymous class assigned to variable without @SuppressWarnings should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.Map;

class NonSerializableKey {}

public class PosCase6 implements Serializable {
    private Map<NonSerializableKey, String> map; // [REPORTED LINE]
    
    public void method() {
        // Anonymous class assigned to variable without @SuppressWarnings("serial")
        Serializable anonymous = new Serializable() {};
    }
}
```

### `PosCase7.java`
Source: `output/codeql/NonSerializableField/src/main/java/scensct/core/pos/PosCase7.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class via inheritance with Map containing non-serializable key type and anonymous class not assigned to variable should be flagged.

```java
// Serializable class via inheritance with Map containing non-serializable key type and anonymous class not assigned to variable should be flagged.
package scensct.core.pos;

import java.io.Serializable;
import java.util.Map;

class NonSerializableKey2 {}

public class PosCase7 implements Serializable {
    private Map<NonSerializableKey2, Serializable> map; // [REPORTED LINE]
    
    public void method() {
        // Anonymous class not assigned to a variable
        new Serializable() {};
    }
}
```

### `PosCase2.java`
Source: `output/sonarqube/SerializableFieldInSerializableClassCheck/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Serializable class with non-serializable Object field without initializer should be flagged as violation.

```java
// Serializable class with non-serializable Object field without initializer should be flagged as violation.
package scensct.core.pos;

import java.io.Serializable;

public class PosCase2 implements Serializable {
    // Non-static, non-transient field of non-serializable type (Object) with no initializer
    public Object obj; // [REPORTED LINE]
}
```

## Cause analysis
The rule is documented as follows: Detects non-serializable fields in Serializable classes that are not handled by custom serialization methods or serialPersistentFields.

In this set the problem shows up around suppression or annotations. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/errorprone.xml/NonSerializableClass`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by zbynek** [comment]
> The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. T

Repeating the same AI-generated phrases over and over again makes these reports really hard to read. Also the code in the samples contains a lot of unrelated and misleading bits.

I summarized the actual issues reported here in https://github.com/pmd/pmd/issues/6547
